File: src/timing.cpp
#include "ccextractor.h"

/* Provide the current time since the file (or the first file) started
 * in ms using PTS time information.
 * Requires: frames_since_ref_time, current_tref
 */

// int ignore_fts=1; // Don't do any kind of timing stuff. Assume it's being done externally (as happens in MP4)

void set_fts(void)
{
    int pts_jump = 0;

    // ES don't have PTS unless GOP timing is used
    if (!pts_set && stream_mode==CCX_SM_ELEMENTARY_OR_NOT_FOUND)
        return;

    // First check for timeline jump (only when min_pts was
    // set (implies sync_pts).
    int dif = 0;
    if (pts_set == 2)
    {
        dif=(int) (current_pts-sync_pts)/MPEG_CLOCK_FREQ;

        // Used to distinguish gaps with missing caption information from
        // jumps in the timeline.  (Currently only used for dvr-ms/NTSC
        // recordings)
        if ( CaptionGap )
            dif = 0;

        // Disable sync check for raw formats - they have the right timeline.
        // Also true for bin formats, but -nosync might have created a
        // broken timeline for debug purposes.
		// Disable too in MP4, specs doesn't say that there can't be a jump
        switch (stream_mode)
        {
            case CCX_SM_MCPOODLESRAW:
            case CCX_SM_RCWT:
			case CCX_SM_MP4:
			case CCX_SM_HEX_DUMP:
                dif = 0;
                break;
            default:
                break;
        }

        if (dif < -0.2 || dif >=5 )
        {
            // ATSC specs: More than 3501 ms means missing component
            mprint ("\nWarning: Reference clock has changed abruptly (%d seconds filepos=%lld), attempting to synchronize\n", (int) dif, past);
            mprint ("Last sync PTS value: %lld\n",sync_pts);
            mprint ("Current PTS value: %lld\n",current_pts);
            pts_jump = 1;
            pts_big_change=1;

            // Discard the gap if it is not on an I-frame or temporal reference
            // zero.
            if(current_tref != 0 && current_picture_coding_type != CCX_FRAME_TYPE_I_FRAME)
            {
                fts_now = fts_max;
                mprint ("Change did not occur on first frame - probably a broken GOP\n");
                return;
            }
        }
    }

    // Set min_pts, fts_offset
    if (pts_set!=0)
    {
        pts_set=2;

        // Use this part only the first time min_pts is set. Later treat
        // it as a reference clock change
        if (current_pts<min_pts && !pts_jump)
        {
            // If this is the first GOP, and seq 0 was not encountered yet
            // we might reset min_pts/fts_offset again

            min_pts=current_pts;

            // Avoid next async test
            sync_pts = LLONG(current_pts
                             -current_tref*1000.0/current_fps
                             *(MPEG_CLOCK_FREQ/1000));

            if(current_tref == 0)
            {   // Earliest time in GOP.
                fts_offset = 0;
            }
            else if ( total_frames_count-frames_since_ref_time == 0 )
            {   // If this is the first frame (PES) there cannot be an offset.
                // This part is also reached for dvr-ms/NTSC (RAW) as 
                // total_frames_count = frames_since_ref_time = 0 when
                // this is called for the first time.
                fts_offset = 0;
            }
            else
            {   // It needs to be "+1" because the current frame is
                // not yet counted.
                fts_offset = LLONG((total_frames_count
                                    -frames_since_ref_time+1)
                                   *1000.0/current_fps);
            }
            dbg_print(CCX_DMT_TIME, "\nFirst sync time    PTS: %s %+lldms (time before this PTS)\n",
                   print_mstime(min_pts/(MPEG_CLOCK_FREQ/1000)),
                   fts_offset );
            dbg_print(CCX_DMT_TIME, "Total_frames_count %u frames_since_ref_time %u\n",
                   total_frames_count, frames_since_ref_time);
        }

        // -nosync diasbles syncing
        if (pts_jump && !ccx_options.nosync)
        {
            // The current time in the old time base is calculated using
            // sync_pts (set at the beginning of the last GOP) plus the
            // time of the frames since then.
            fts_offset = fts_offset
                + (sync_pts-min_pts)/(MPEG_CLOCK_FREQ/1000)
                + LLONG(frames_since_ref_time*1000/current_fps);
            fts_max = fts_offset;

            // Start counting again from here
            pts_set=1; // Force min to be set again

            // Avoid next async test - the gap might have occured on
            // current_tref != 0.
            sync_pts = LLONG(current_pts
                             -current_tref*1000.0/current_fps
                             *(MPEG_CLOCK_FREQ/1000));
            // Set min_pts = sync_pts as this is used for fts_now
            min_pts = sync_pts;

            dbg_print(CCX_DMT_TIME, "\nNew min PTS time: %s %+lldms (time before this PTS)\n",
                   print_mstime(min_pts/(MPEG_CLOCK_FREQ/1000)),
                   fts_offset );
        }
    }

    // Set sync_pts, fts_offset
    if(current_tref == 0)
        sync_pts = current_pts;

    // Reset counters
    cb_field1 = 0;
    cb_field2 = 0;
    cb_708 = 0;

    // Avoid wrong "Calc. difference" and "Asynchronous by" numbers
    // for uninitialized min_pts
	if (1) // CFS: Remove or think decent condition
	{
		if ( pts_set )
		{
			// If pts_set is TRUE we have min_pts
			fts_now = LLONG((current_pts-min_pts)/(MPEG_CLOCK_FREQ/1000)
							+ fts_offset);
		}
		else
		{
			// No PTS info at all!!
			fatal(EXIT_BUG_BUG,
				  "No PTS info. Please write bug report.");
		}
	}
    if ( fts_now > fts_max )
    {
        fts_max = fts_now;
    }
}


LLONG get_fts(void)
{
    LLONG fts;

    switch (current_field)
    {
        case 1:
            fts = fts_now + fts_global + cb_field1*1001/30;
            break;
        case 2:
            fts = fts_now + fts_global + cb_field2*1001/30;
            break;
        case 3:
            fts = fts_now + fts_global + cb_708*1001/30;
            break;
        default:
            fatal(EXIT_BUG_BUG, "Cannot be reached!");
    }

    return fts;
}

LLONG get_fts_max(void)
{
    // This returns the maximum FTS that belonged to a frame.  Caption block
    // counters are not applicable.
    return fts_max + fts_global;
}

/* Fill a static buffer with a time string (hh:mm:ss:ms) corresponding
to the microsecond value in mstime. */
char *print_mstime2buf( LLONG mstime , char *buf )
{
    unsigned hh,mm,ss,ms;
    int signoffset = (mstime < 0 ? 1 : 0);

    if (mstime<0) // Avoid loss of data warning with abs()
        mstime=-mstime;     
    hh = (unsigned) (mstime/1000/60/60);
    mm = (unsigned) (mstime/1000/60 - 60*hh);
    ss = (unsigned) (mstime/1000 - 60*(mm + 60*hh));
    ms = (int) (mstime - 1000*(ss + 60*(mm + 60*hh)));

    buf[0]='-';
    sprintf (buf+signoffset, "%02u:%02u:%02u:%03u",hh,mm,ss,ms);

    return buf;
}

/* Fill a static buffer with a time string (hh:mm:ss:ms) corresponding
to the microsecond value in mstime. */
char *print_mstime( LLONG mstime )
{
    static char buf[15]; // 14 should be long enough
	return print_mstime2buf (mstime, buf);
}

/* Helper function for to display debug timing info. */
void print_debug_timing( void )
{
    // Avoid wrong "Calc. difference" and "Asynchronous by" numbers
    // for uninitialized min_pts
    LLONG tempmin_pts = (min_pts==0x01FFFFFFFFLL ? sync_pts : min_pts);

    mprint("Sync time stamps:  PTS: %s                ",
           print_mstime((sync_pts)/(MPEG_CLOCK_FREQ/1000)) );
    mprint("GOP: %s      \n", print_mstime(gop_time.ms));

    // Length first GOP to last GOP
    LLONG goplenms = LLONG(gop_time.ms - first_gop_time.ms);
    // Length at last sync point
    LLONG ptslenms = unsigned((sync_pts-tempmin_pts)/(MPEG_CLOCK_FREQ/1000)
                              + fts_offset);

    mprint("Last               FTS: %s",
           print_mstime(get_fts_max()));
    mprint("      GOP start FTS: %s\n",
           print_mstime(fts_at_gop_start));

    // Times are based on last GOP and/or sync time
    mprint("Max FTS diff. to   PTS:       %6lldms              GOP:       %6lldms\n\n",
           get_fts_max()+LLONG(1000.0/current_fps)-ptslenms,
           get_fts_max()+LLONG(1000.0/current_fps)-goplenms);
}

void calculate_ms_gop_time (struct gop_time_code *g)
{
    int seconds=(g->time_code_hours*3600)+(g->time_code_minutes*60)+g->time_code_seconds;
    g->ms = LLONG( 1000*(seconds + g->time_code_pictures/current_fps) );
    if (gop_rollover)
        g->ms += 24*60*60*1000;
}

int gop_accepted(struct gop_time_code* g )
{
    if (! ((g->time_code_hours <= 23) 
        && (g->time_code_minutes <= 59) 
        && (g->time_code_seconds <= 59) 
        && (g->time_code_pictures <= 59)))
        return 0;

    if (gop_time.time_code_hours==23 && gop_time.time_code_minutes==59 &&
        g->time_code_hours==0 && g->time_code_minutes==0)
    {
        gop_rollover = 1;
        return 1;
    }
    if (gop_time.inited)
    {
        if (gop_time.ms > g->ms)
        {
            // We are going back in time but it's not a complete day rollover
            return 0;
        }
    }
    return 1;
}

File: src/rust/lib_ccxr/src/util/time/timing.rs
use crate::common::FrameType;
use crate::util::log::{debug, info, DebugMessageFlag};
use crate::util::time::{FrameCount, GopTimeCode, MpegClockTick, Timestamp};
use std::sync::RwLock;

/// The maximum allowed difference between [`TimingContext::current_pts`] and [`TimingContext::sync_pts`] in seconds.
///
/// If the difference crosses this value, a PTS jump has occured and is handled accordingly.
const MAX_DIF: i64 = 5;

/// A unique global instance of [`GlobalTimingInfo`] to be used throughout the program.
pub static GLOBAL_TIMING_INFO: RwLock<GlobalTimingInfo> = RwLock::new(GlobalTimingInfo::new());

const DEFAULT_FRAME_RATE: f64 = 30000.0 / 1001.0;

/// Represents the status of [`TimingContext::current_pts`] and [`TimingContext::min_pts`]
#[derive(Copy, Clone, Debug, Eq, PartialEq)]
pub enum PtsSet {
    No,
    Received,
    MinPtsSet,
}

/// Represent the field of 608 or 708 caption.
#[derive(Copy, Clone, Debug, Eq, PartialEq)]
pub enum CaptionField {
    Field1,
    Field2,
    Cea708,
}

/// A collective struct for storing timing-related information when decoding a file.
///
/// [`GlobalTimingInfo`] serves a similar purpose. The only difference is that its lifetime is
/// global.
#[derive(Debug)]
pub struct TimingContext {
    pub pts_set: PtsSet,
    /// if true then don't adjust again.
    min_pts_adjusted: bool,
    pub current_pts: MpegClockTick,
    pub current_picture_coding_type: FrameType,
    /// Store temporal reference of current frame.
    pub current_tref: FrameCount,
    pub min_pts: MpegClockTick,
    pub sync_pts: MpegClockTick,
    /// No screen should start before this FTS
    pub minimum_fts: Timestamp,
    /// Time stamp of current file (w/ fts_offset, w/o fts_global).
    pub fts_now: Timestamp,
    /// Time before first sync_pts.
    pub fts_offset: Timestamp,
    /// Time before first GOP.
    pub fts_fc_offset: Timestamp,
    /// Remember the maximum fts that we saw in current file.
    pub fts_max: Timestamp,
    /// Duration of previous files (-ve mode).
    pub fts_global: Timestamp,
    pub sync_pts2fts_set: bool,
    pub sync_pts2fts_fts: Timestamp,
    pub sync_pts2fts_pts: MpegClockTick,
    /// PTS resets when current_pts is lower than prev.
    pts_reset: bool,
}

/// Settings for overall timing functionality in [`TimingContext`].
#[derive(Debug)]
pub struct TimingSettings {
    /// If true, timeline jumps will be ignored. This is important in several input formats that are assumed to have correct timing, no matter what.
    pub disable_sync_check: bool,

    /// If true, there will be no sync at all. Mostly useful for debugging.
    pub no_sync: bool,

    // Needs to be set, as it's used in set_fts.
    pub is_elementary_stream: bool,
}

/// A collective struct to store global timing-related information while decoding a file.
///
/// [`TimingContext`] serves a similar purpose. The only difference is that its lifetime is not
/// global and its information could be reset while execution of program.
#[derive(Debug)]
pub struct GlobalTimingInfo {
    // Count 608 (per field) and 708 blocks since last set_fts() call
    pub cb_field1: u64,
    pub cb_field2: u64,
    pub cb_708: u64,
    pub pts_big_change: bool,
    pub current_fps: f64,
    pub frames_since_ref_time: FrameCount,
    pub total_frames_count: FrameCount,
    pub gop_time: Option<GopTimeCode>,
    pub first_gop_time: Option<GopTimeCode>,
    pub fts_at_gop_start: Timestamp,
    pub gop_rollover: bool,
    pub timing_settings: TimingSettings,
}

impl TimingContext {
    /// Create a new [`TimingContext`].
    pub fn new() -> TimingContext {
        TimingContext {
            pts_set: PtsSet::No,
            min_pts_adjusted: false,
            current_pts: MpegClockTick::new(0),
            current_picture_coding_type: FrameType::ResetOrUnknown,
            current_tref: FrameCount::new(0),
            min_pts: MpegClockTick::new(0x01FFFFFFFF),
            sync_pts: MpegClockTick::new(0),
            minimum_fts: Timestamp::from_millis(0),
            fts_now: Timestamp::from_millis(0),
            fts_offset: Timestamp::from_millis(0),
            fts_fc_offset: Timestamp::from_millis(0),
            fts_max: Timestamp::from_millis(0),
            fts_global: Timestamp::from_millis(0),
            sync_pts2fts_set: false,
            sync_pts2fts_fts: Timestamp::from_millis(0),
            sync_pts2fts_pts: MpegClockTick::new(0),
            pts_reset: false,
        }
    }

    /// Add `pts` to `TimingContext::current_pts`.
    ///
    /// It also checks for PTS resets.
    pub fn add_current_pts(&mut self, pts: MpegClockTick) {
        self.set_current_pts(self.current_pts + pts)
    }

    /// Set `TimingContext::current_pts` to `pts`.
    ///
    /// It also checks for PTS resets.
    pub fn set_current_pts(&mut self, pts: MpegClockTick) {
        let prev_pts = self.current_pts;
        self.current_pts = pts;
        if self.pts_set == PtsSet::No {
            self.pts_set = PtsSet::Received
        }
        debug!(msg_type = DebugMessageFlag::VIDEO_STREAM; "PTS: {} ({:8})", self.current_pts.as_timestamp().to_hms_millis_time(':').unwrap(), self.current_pts.as_i64());
        debug!(msg_type = DebugMessageFlag::VIDEO_STREAM; "  FTS: {} \n", self.fts_now.to_hms_millis_time(':').unwrap());

        // Check if PTS reset
        if self.current_pts < prev_pts {
            self.pts_reset = true;
        }
    }

    pub fn set_fts(&mut self) -> bool {
        let mut timing_info = GLOBAL_TIMING_INFO.write().unwrap();

        let mut pts_jump = false;

        // ES don't have PTS unless GOP timing is used
        if self.pts_set == PtsSet::No && timing_info.timing_settings.is_elementary_stream {
            return true;
        }

        // First check for timeline jump (only when min_pts was set (implies sync_pts)).
        if self.pts_set == PtsSet::MinPtsSet {
            let dif = if timing_info.timing_settings.disable_sync_check {
                // Disables sync check. Used for several input formats.
                0
            } else {
                (self.current_pts - self.sync_pts).as_timestamp().seconds()
            };

            // This was -0.2 before. TODO: find out why, since dif is integer?
            if !(0..MAX_DIF).contains(&dif) {
                // ATSC specs: More than 3501 ms means missing component
                info!("\nWarning: Reference clock has changed abruptly ({} seconds filepos={}), attempting to synchronize\n", dif, "Unable to get file position"); // TODO: get the file position somehow
                info!("Last sync PTS value: {}\n", self.sync_pts.as_i64());
                info!("Current PTS value: {}\n", self.current_pts.as_i64());
                info!("Note: You can disable this behavior by adding -ignoreptsjumps to the parameters.\n");

                pts_jump = true;
                timing_info.pts_big_change = true;

                // Discard the gap if it is not on an I-frame or temporal reference zero.
                if self.current_tref.as_u64() != 0
                    && self.current_picture_coding_type != FrameType::IFrame
                {
                    self.fts_now = self.fts_max;
                    info!("Change did not occur on first frame - probably a broken GOP\n");
                    return true;
                }
            }
        }

        // If min_pts was set just before a rollover we compensate by "roll-oving" it too
        if self.pts_set == PtsSet::MinPtsSet && !self.min_pts_adjusted {
            // min_pts set
            // We want to be aware of the upcoming rollover, not after it happened, so we don't take
            // the 3 most significant bits but the 3 next ones
            let min_pts_big_bits = (self.min_pts.as_i64() >> 30) & 7;
            let cur_pts_big_bits = (self.current_pts.as_i64() >> 30) & 7;
            if cur_pts_big_bits == 7 && min_pts_big_bits == 0 {
                // Huge difference possibly means the first min_pts was actually just over the boundary
                // Take the current_pts (smaller, accounting for the rollover) instead
                self.min_pts = self.current_pts;
                self.min_pts_adjusted = true;
            } else if (1..=6).contains(&cur_pts_big_bits) {
                // Far enough from the boundary
                // Prevent the eventual difference with min_pts to make a bad adjustment
                self.min_pts_adjusted = true;
            }
        }

        // Set min_pts, fts_offset
        if self.pts_set != PtsSet::No {
            self.pts_set = PtsSet::MinPtsSet;

            // Use this part only the first time min_pts is set. Later treat
            // it as a reference clock change
            if self.current_pts < self.min_pts && !pts_jump {
                // If this is the first GOP, and seq 0 was not encountered yet
                // we might reset min_pts/fts_offset again

                self.min_pts = self.current_pts;

                // Avoid next async test
                self.sync_pts = self.current_pts
                    - self
                        .current_tref
                        .as_mpeg_clock_tick(timing_info.current_fps);

                if self.current_tref.as_u64() == 0
                    || (timing_info.total_frames_count - timing_info.frames_since_ref_time).as_u64()
                        == 0
                {
                    // Earliest time in GOP.
                    // OR
                    // If this is the first frame (PES) there cannot be an offset.
                    // This part is also reached for dvr-ms/NTSC (RAW) as
                    // total_frames_count = frames_since_ref_time = 0 when
                    // this is called for the first time.
                    self.fts_offset = Timestamp::from_millis(0);
                } else {
                    // It needs to be "+1" because the current frame is
                    // not yet counted.
                    let one_frame = FrameCount::new(1);
                    self.fts_offset = (timing_info.total_frames_count
                        - timing_info.frames_since_ref_time
                        + one_frame)
                        .as_timestamp(timing_info.current_fps);
                }
                debug!(
                    msg_type = DebugMessageFlag::TIME;
                    "\nFirst sync time    PTS: {} {:+}ms (time before this PTS)\n",
                    self.min_pts.as_timestamp().to_hms_millis_time(':').unwrap(),
                    self.fts_offset.millis()
                );
                debug!(
                    msg_type = DebugMessageFlag::TIME;
                    "Total_frames_count {} frames_since_ref_time {}\n",
                    timing_info.total_frames_count.as_u64(),
                    timing_info.frames_since_ref_time.as_u64()
                );
            }

            // -nosync disables syncing
            if pts_jump && !timing_info.timing_settings.no_sync {
                // The current time in the old time base is calculated using
                // sync_pts (set at the beginning of the last GOP) plus the
                // time of the frames since then.
                self.fts_offset = self.fts_offset
                    + (self.sync_pts - self.min_pts).as_timestamp()
                    + timing_info
                        .frames_since_ref_time
                        .as_timestamp(timing_info.current_fps);
                self.fts_max = self.fts_offset;

                // Start counting again from here
                self.pts_set = PtsSet::Received; // Force min to be set again
                self.sync_pts2fts_set = false; // Make note of the new conversion values

                // Avoid next async test - the gap might have occured on
                // current_tref != 0.
                self.sync_pts = self.current_pts
                    - self
                        .current_tref
                        .as_mpeg_clock_tick(timing_info.current_fps);
                // Set min_pts = sync_pts as this is used for fts_now
                self.min_pts = self.sync_pts;

                debug!(
                    msg_type = DebugMessageFlag::TIME;
                    "\nNew min PTS time: {} {:+}ms (time before this PTS)\n",
                    self.min_pts.as_timestamp().to_hms_millis_time(':').unwrap(),
                    self.fts_offset.millis()
                );
            }
        }

        // Set sync_pts, fts_offset
        if self.current_tref.as_u64() == 0 {
            self.sync_pts = self.current_pts;
        }

        // Reset counters
        timing_info.cb_field1 = 0;
        timing_info.cb_field2 = 0;
        timing_info.cb_708 = 0;

        // Avoid wrong "Calc. difference" and "Asynchronous by" numbers
        // for uninitialized min_pts
        if true {
            // CFS: Remove or think decent condition
            if self.pts_set != PtsSet::No {
                // If pts_set is TRUE we have min_pts
                self.fts_now = (self.current_pts - self.min_pts).as_timestamp() + self.fts_offset;
                if !self.sync_pts2fts_set {
                    self.sync_pts2fts_pts = self.current_pts;
                    self.sync_pts2fts_fts = self.fts_now;
                    self.sync_pts2fts_set = true;
                }
            } else {
                // No PTS info at all!!
                info!("Set PTS called without any global timestamp set\n");
                return false;
            }
        }
        if self.fts_now > self.fts_max {
            self.fts_max = self.fts_now;
        }

        // If PTS resets, then fix minimum_fts and fts_max
        if self.pts_reset {
            self.minimum_fts = Timestamp::from_millis(0);
            self.fts_max = self.fts_now;
            self.pts_reset = false;
        }

        true
    }

    /// Returns the total FTS.
    ///
    /// Caption block counters are included.
    pub fn get_fts(&self, current_field: CaptionField) -> Timestamp {
        let timing_info = GLOBAL_TIMING_INFO.read().unwrap();
        let count = match current_field {
            CaptionField::Field1 => timing_info.cb_field1,
            CaptionField::Field2 => timing_info.cb_field2,
            CaptionField::Cea708 => timing_info.cb_708,
        };
        self.fts_now + self.fts_global + Timestamp::from_millis((count as i64) * 1001 / 30)
    }

    /// This returns the maximum FTS that belonged to a frame.
    ///
    /// Caption block counters are not applicable.
    pub fn get_fts_max(&self) -> Timestamp {
        self.fts_max + self.fts_global
    }

    /// Log FTS and PTS information for debugging purpose.
    pub fn print_debug_timing(&self) {
        let zero = Timestamp::from_millis(0);
        let timing_info = GLOBAL_TIMING_INFO.read().unwrap();

        let gop_time = timing_info.gop_time.map(|x| x.timestamp()).unwrap_or(zero);
        let first_gop_time = timing_info
            .first_gop_time
            .map(|x| x.timestamp())
            .unwrap_or(zero);

        // Avoid wrong "Calc. difference" and "Asynchronous by" numbers
        // for uninitialized min_pts
        let tempmin_pts = if self.min_pts.as_i64() == 0x01FFFFFFFF {
            self.sync_pts
        } else {
            self.min_pts
        };

        info!(
            "Sync time stamps:  PTS: {}                ",
            self.sync_pts
                .as_timestamp()
                .to_hms_millis_time(':')
                .unwrap()
        );
        info!(
            "      GOP start FTS: {}\n",
            gop_time.to_hms_millis_time(':').unwrap()
        );

        // Length first GOP to last GOP
        let goplenms = gop_time - first_gop_time;
        // Length at last sync point
        let ptslenms = (self.sync_pts - tempmin_pts).as_timestamp() + self.fts_offset;

        info!(
            "Last               FTS: {}",
            self.get_fts_max().to_hms_millis_time(':').unwrap()
        );
        info!(
            "      GOP start FTS: {}\n",
            timing_info
                .fts_at_gop_start
                .to_hms_millis_time(':')
                .unwrap()
        );

        let one_frame = FrameCount::new(1).as_timestamp(timing_info.current_fps);

        // Times are based on last GOP and/or sync time
        info!(
            "Max FTS diff. to   PTS:       {:6}ms              GOP:       {:6}ms\n\n",
            (self.get_fts_max() + one_frame - ptslenms)
                .to_hms_millis_time(':')
                .unwrap(),
            (self.get_fts_max() + one_frame - goplenms)
                .to_hms_millis_time(':')
                .unwrap()
        );
    }

    /// Constructs a [`TimingContext`] from its individual fields.
    ///
    /// # Safety
    ///
    /// Make sure that [`TimingContext`] is in a valid state.
    #[allow(clippy::too_many_arguments)]
    pub unsafe fn from_raw_parts(
        pts_set: PtsSet,
        min_pts_adjusted: bool,
        current_pts: MpegClockTick,
        current_picture_coding_type: FrameType,
        current_tref: FrameCount,
        min_pts: MpegClockTick,
        sync_pts: MpegClockTick,
        minimum_fts: Timestamp,
        fts_now: Timestamp,
        fts_offset: Timestamp,
        fts_fc_offset: Timestamp,
        fts_max: Timestamp,
        fts_global: Timestamp,
        sync_pts2fts_set: bool,
        sync_pts2fts_fts: Timestamp,
        sync_pts2fts_pts: MpegClockTick,
        pts_reset: bool,
    ) -> TimingContext {
        TimingContext {
            pts_set,
            min_pts_adjusted,
            current_pts,
            current_picture_coding_type,
            current_tref,
            min_pts,
            sync_pts,
            minimum_fts,
            fts_now,
            fts_offset,
            fts_fc_offset,
            fts_max,
            fts_global,
            sync_pts2fts_set,
            sync_pts2fts_fts,
            sync_pts2fts_pts,
            pts_reset,
        }
    }

    /// Returns the individual fields of a [`TimingContext`].
    ///
    /// # Safety
    ///
    /// Certain fields are supposed to be private.
    #[allow(clippy::type_complexity)]
    pub unsafe fn as_raw_parts(
        &self,
    ) -> (
        PtsSet,
        bool,
        MpegClockTick,
        FrameType,
        FrameCount,
        MpegClockTick,
        MpegClockTick,
        Timestamp,
        Timestamp,
        Timestamp,
        Timestamp,
        Timestamp,
        Timestamp,
        bool,
        Timestamp,
        MpegClockTick,
        bool,
    ) {
        let TimingContext {
            pts_set,
            min_pts_adjusted,
            current_pts,
            current_picture_coding_type,
            current_tref,
            min_pts,
            sync_pts,
            minimum_fts,
            fts_now,
            fts_offset,
            fts_fc_offset,
            fts_max,
            fts_global,
            sync_pts2fts_set,
            sync_pts2fts_fts,
            sync_pts2fts_pts,
            pts_reset,
        } = *self;

        (
            pts_set,
            min_pts_adjusted,
            current_pts,
            current_picture_coding_type,
            current_tref,
            min_pts,
            sync_pts,
            minimum_fts,
            fts_now,
            fts_offset,
            fts_fc_offset,
            fts_max,
            fts_global,
            sync_pts2fts_set,
            sync_pts2fts_fts,
            sync_pts2fts_pts,
            pts_reset,
        )
    }
}

impl GlobalTimingInfo {
    /// Create a new instance of [`GlobalTimingInfo`].
    const fn new() -> GlobalTimingInfo {
        GlobalTimingInfo {
            cb_field1: 0,
            cb_field2: 0,
            cb_708: 0,
            pts_big_change: false,
            current_fps: DEFAULT_FRAME_RATE, // 29.97
            // TODO: Get from framerates_values[] instead
            frames_since_ref_time: FrameCount::new(0),
            total_frames_count: FrameCount::new(0),
            gop_time: None,
            first_gop_time: None,
            fts_at_gop_start: Timestamp::from_millis(0),
            gop_rollover: false,
            timing_settings: TimingSettings {
                disable_sync_check: false,
                no_sync: false,
                is_elementary_stream: false,
            },
        }
    }
}

impl Default for TimingContext {
    fn default() -> Self {
        Self::new()
    }
}
